### ISSUE TITLE
Crashes in prompt cause a repaint loop

### ISSUE THREAD

**Issue by faho** [fish-shell/fish-shell#9796]
`fish_job_summary` calls `commandline -f repaint`, which, if it is called via something in fish_prompt, causes it to loop.

I found this when I noticed that fossil crashes if $PWD doesn't exist, so the test case is:

```fish
function fish_prompt; fossil branch current; end
mkdir foo
cd foo
rm -rf $PWD
```

It'll end up printing a constant stream of

> fish: Job 1, 'fossil branch current' terminated by signal SIGABRT (Abort)

until you press ctrl-c. While this should be a bug in fossil, the general principle stands.

I'm not sure how to detect this and turn fish_job_summary off.

**Comment by lavafroth** [comment]
Hi. I have been looking into this but I can't seem to reproduce the bug. Perhaps this has been fixed?

**Comment by faho** [comment]
Yup, the repaint was incidentally removed in 9158395d10cfa584a988ba093cf7a93a70b48578

**Comment by faho** [comment]
Ah, turns out testing with SIGKILL doesn't work.

I had used `bash -c 'kill -9 $$'`, which doesn't trigger fish_job_summary. It does with `kill -6`.

**Comment by ridiculousfish** [comment]
The looping is because of this:

```rust
    fn prepare_to_select(&mut self) {
        // Fire any pending events and reap stray processes, including printing exit status messages.
        event::fire_delayed(self.parser);
        if job_reap(self.parser, true) {
            reader_schedule_prompt_repaint();
        }
    }
```

that is, right before we wait for user input, we reap processes and allow printing messages; if a message is printed we schedule a prompt repaint and then react to that instead of select:

```rust
            // We are going to block; but first allow any override to inject events.
            self.prepare_to_select();
            if let Some(mevt) = self.try_pop() {
                return Some(mevt);
            }
```

still unclear what the right approach is. It's worth noting that this issue seems to have existed forever; still if we can fix it we should since it's a very bad state to get into.

**Comment by krobelus** [comment]
Ah so we never even read stdin.
We could defer reaping jobs from prompt generation X until the current prompt generation is X+1.
This change need only apply to jobs spawned within `exec_prompt`.
That should slow down the loop to only move one cycle whenever there is a legit repaint.

I haven't explored all alternatives, one of them would be an interrupt key that we always prioritize over all other events (possibly such keys could replace some signal-handling logic)

**Comment by ridiculousfish** [comment]
I think I have a pretty solid fix (one liner!) in #10891
